Adding aws_acm_export_certificates

diodonfrost opened the pull request:
##### SUMMARY

PR to introduce aws_acm_export_certificates module.

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

aws_acm_export_certificates 

##### ADDITIONAL INFORMATION

This module allow to export [private certificates](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/acm.html#ACM.Client.export_certificate) from aws certificates manager via the Ansible framework.

diodonfrost committed:
Adding aws_acm_export_certificates

diodonfrost committed:
add new line at the end of file

willthames commented on lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py:
```diff
@@ -0,0 +1,138 @@
+        )
+    )
+
+    module = AnsibleModule(argument_spec=argument_spec)
```
Recommend using AnsibleAWSModule to get all the latest features - can just use `module.client` to get a client, and no need for region checking

Also without it your `module.fail_json_aws` won't work.

diodonfrost replied:
```
-from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.aws.core import AnsibleAWSModule
 from ansible.module_utils.ec2 import (
     boto3_conn,
     ec2_argument_spec,
@@ -76,7 +76,7 @@ except ImportError:
     pass  # caught by imported AnsibleAWSModule

 def main():
@@ -113,24 +112,12 @@ def main():
         )
     )

-    module = AnsibleModule(argument_spec=argument_spec)
+    module = AnsibleAWSModule(argument_spec=argument_spec)
+    connection = module.client('acm')

     if not HAS_BOTO3:
         module.fail_json(msg='boto3 and botocore are required.')

-    region, ec2_url, aws_connect_kwargs = get_aws_connection_info(module, boto3=True)
-    if region:
-        connection = boto3_conn(
-            module,
-            conn_type='client',
-            resource='acm',
-            region=region,
-            endpoint=ec2_url,
-            **aws_connect_kwargs
-        )
-    else:
-        module.fail_json(msg="region must be specified")
-
     export_acm_certificates(connection, module)
````
I used AnsibleAWSModule and `module.client`

willthames commented on lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py:
```diff
@@ -0,0 +1,138 @@
+    pass  # caught by imported AnsibleAWSModule
+
+
+def normalize_certificates(acm_certificates):
```
This function could probably just be replaced by `camel_dict_to_snake_dict(acm_certificates)`

diodonfrost replied:
```
-def normalize_certificates(acm_certificates):
+def camel_dict_to_snake_dict(acm_certificates):
```
Use `camel_dict_to_snake_dict`

willthames commented on lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py:
```diff
@@ -0,0 +1,138 @@
+        module.fail_json_aws(e)
+
+    acm_certificates = normalize_certificates(response)
+    module.exit_json(certificates=acm_certificates)
```
This should be singular

diodonfrost replied:
```
-    acm_certificates = normalize_certificates(response)
-    module.exit_json(certificates=acm_certificates)
+    module.exit_json(certificates=camel_dict_to_snake_dict(response))
```

diodonfrost committed:
Fix linting issues

willthames commented on lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py:
```diff
@@ -0,0 +1,125 @@
+    module = AnsibleAWSModule(argument_spec=argument_spec)
+    connection = module.client('acm')
+
+    if not HAS_BOTO3:
```
Not required with AnsibleAWSModule

diodonfrost replied:
```
-    if not HAS_BOTO3:
-        module.fail_json(msg='boto3 and botocore are required.')
-
```

Removing boto3 check

willthames commented on lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py:
```diff
@@ -0,0 +1,125 @@
+    argument_spec = ec2_argument_spec()
+    argument_spec.update(
+        dict(
+            certificate_arn=dict(type='str', required=True),
```
`type='str'` is the default, so not needed here.

diodonfrost replied:
```
@@ -107,17 +97,14 @@ def main():
     argument_spec = ec2_argument_spec()
     argument_spec.update(
         dict(
-            certificate_arn=dict(type='str', required=True),
-            passphrase=dict(type='str', required=True)
+            certificate_arn=dict(required=True),
+            passphrase=dict(required=True)
         )
     )
```
Removing `type='str`

willthames commented on lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py:
```diff
@@ -0,0 +1,125 @@
+
+
+from ansible.module_utils.aws.core import AnsibleAWSModule
+from ansible.module_utils.ec2 import (
```
These imports aren't needed but you will need `camel_dict_to_snake_dict` from `ansible.module_utils.ec2`

diodonfrost replied:
```
-def camel_dict_to_snake_dict(acm_certificates):                       
-    certificate = acm_certificates.get('Certificate', None)           
-    certificate_chain = acm_certificates.get('CertificateChain', None)
-    private_key = acm_certificates.get('PrivateKey', None)            
-    return {                                                          
-        'certificate': certificate,                                   
-        'certificate_chain': certificate_chain,                       
-        'private_key': private_key                                    
-    }                                                                 
-                                                                      
-                                                                      
 def export_acm_certificates(connection, module):                      
     params = {                                                        
```
Removing `camel_dict_to_snake_dict` function

willthames commented on lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py:
```diff
@@ -0,0 +1,125 @@
+'''
+
+RETURN = '''
+certificates:
```
This is still only one certificate. The name of the module, and the return value should be changed.

diodonfrost replied:
```
-certificates:
-    description: The list of certificates and private key returned by the AWS Certificate Manager
+export:
+    description: Export the certificate, the certificate chain, and the encrypted private key associated with the public key embedded in the certificate
     returned: always
     type: dict
     sample:
```
Change ansible return value.

willthames commented on lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py:
```diff
@@ -0,0 +1,125 @@
+    pass  # caught by imported AnsibleAWSModule
+
+
+def camel_dict_to_snake_dict(acm_certificates):
```
Sorry, use `camel_dict_to_snake_dict` from `ansible.module_utils.ec2`

diodonfrost replied:
```
     try:
         response = connection.export_certificate(**kwargs)
-    except (ClientError, ParamValidationError) as e:
-        module.fail_json_aws(e)
+    except (ClientError, ParamValidationError) as error:
+        module.fail_json(msg="Couldn't obtain private certificates",
+                         exception=traceback.format_exc(),
+                         **camel_dict_to_snake_dict(error.response))

     module.exit_json(certificates=camel_dict_to_snake_dict(response))
````

Use `camel_dict_to_snake_dict` from `ansible.module_utils.ec2`

willthames commented on lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py:
```diff
@@ -0,0 +1,125 @@
+    try:
+        response = connection.export_certificate(**kwargs)
+    except (ClientError, ParamValidationError) as e:
+        module.fail_json_aws(e)
```
Please add a `msg` value here e.g. "Failed to export certificates" just to make it a little clearer

diodonfrost replied:
```
     try:
         response = connection.export_certificate(**kwargs)
-    except (ClientError, ParamValidationError) as e:
-        module.fail_json_aws(e)
+    except (ClientError, ParamValidationError) as error:
+        module.fail_json(msg="Couldn't obtain private certificates",
+                         exception=traceback.format_exc(),
+                         **camel_dict_to_snake_dict(error.response))

     module.exit_json(certificates=camel_dict_to_snake_dict(response))
```
Add msg value during exception

willthames reviewed (changes requested):
Still a few minor tweaks but overall this is looking very good.

willthames commented:
This will also require a test suite (see test/integration/targets for existing ones) - it doesn't have to do much, just enough that we can validate that the module actually works. You may need to use the `command` module with the aws cli to create a certificate as we don't seem to have a module for creating acm certificates.

diodonfrost committed:
fix doc and issues

diodonfrost commented:
AWS acm command [export-certificate](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/acm.html#ACM.Client.export_certificate) only work with [private certificates](https://docs.aws.amazon.com/acm/latest/userguide/gs-acm-request-private.html)

Private certificates can only be requested by [AWS Certificate Manager Private Certificate Authority ](https://docs.aws.amazon.com/acm-pca/latest/userguide/PcaWelcome.html)

I can make a test suite but you need a ACM PCA for it to work.

willthames commented on lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py:
```diff
@@ -0,0 +1,112 @@
+                         exception=traceback.format_exc(),
+                         **camel_dict_to_snake_dict(error.response))
+
+    module.exit_json(certificates=camel_dict_to_snake_dict(response))
```
This doesn't match docs and should still be singular certificate

willthames replied:
Might also be better to just `module.exit_json(**camel_dict_to_snake_dict(response))`

diodonfrost replied:
```
     try:
         response = connection.export_certificate(**kwargs)
     except (ClientError, ParamValidationError) as error:
-        module.fail_json(msg="Couldn't obtain private certificates",
-                         exception=traceback.format_exc(),
-                         **camel_dict_to_snake_dict(error.response))
+        module.fail_json_aws(error, msg="Couldn't obtain private certificate")

-    module.exit_json(certificates=camel_dict_to_snake_dict(response))
+    module.exit_json(**camel_dict_to_snake_dict(response))
```

diodonfrost replied:
refactor `module.exit_json`

willthames commented on lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py:
```diff
@@ -0,0 +1,112 @@
+
+
+def main():
+    argument_spec = ec2_argument_spec()
```
not needed with AnsibleAWSModule

diodonfrost replied:
```
 def main():
-    argument_spec = ec2_argument_spec()
-    argument_spec.update(
-        dict(
-            certificate_arn=dict(required=True),
-            passphrase=dict(required=True)
-        )
+    argument_spec = dict(
+        certificate_arn=dict(required=True),
+        passphrase=dict(required=True)
     )
-
-    module = AnsibleAWSModule(argument_spec=argument_spec)
+
+    module = AnsibleAWSModule(argument_spec=argument_spec, supports_check_mode=False,)
     connection = module.client('acm')
```

diodonfrost replied:
remove `argument_spec = ec2_argument_spec()`

willthames commented on lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py:
```diff
@@ -0,0 +1,112 @@
+    try:
+        response = connection.export_certificate(**kwargs)
+    except (ClientError, ParamValidationError) as error:
+        module.fail_json(msg="Couldn't obtain private certificates",
```
Use module.fail_json_aws(error, msg="Couldn't obtain private certificates")

diodonfrost replied:
```
     try:
         response = connection.export_certificate(**kwargs)
     except (ClientError, ParamValidationError) as error:
-        module.fail_json(msg="Couldn't obtain private certificates",
-                         exception=traceback.format_exc(),
-                         **camel_dict_to_snake_dict(error.response))
+        module.fail_json_aws(error, msg="Couldn't obtain private certificate")
```

Fix `module.fail_json_aws` output

willthames commented on lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py:
```diff
@@ -0,0 +1,112 @@
+# Exporting a private certificate and private key (more details: https://docs.aws.amazon.com/acm/latest/userguide/gs-acm-export-private.html#export-cli)
+
+- name: obtain specific privates certificates
+  aws_acm_export_certificates:
```
Rename the module to be singular as it only exports one certificate

diodonfrost replied:
```
DOCUMENTATION = '''
 ---
-module: aws_acm_export_certificates
+module: aws_acm_export_certificate
 short_description: Exports a private certificate issued by a private certificate authority (CA) for use anywhere
 description:
     - Exports a private certificate issued by a private certificate authority (CA) for use anywhere
@@ -42,22 +42,26 @@ EXAMPLES = '''
 # Note: These examples do not set authentication details, see the AWS Guide for details.
 # Exporting a private certificate and private key (more details: https://docs.aws.amazon.com/acm/latest/userguide/gs-acm-export-private.html#export-cli)

-- name: obtain specific privates certificates
-  aws_acm_export_certificates:
+- name: obtain specific privates certificate
+  aws_acm_export_certificate:
     certificate_arn: "arn:aws:acm:us-east-1:123456789012:certificate/12345678-1234-1234-1234-123456789012"
     passphrase: secretpassword
-  register: acm_certificates
+  register: acm_certificate
```

diodonfrost replied:
rename module to `aws_acm_export_certificate`

willthames commented on lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py:
```diff
@@ -0,0 +1,112 @@
+'''
+
+RETURN = '''
+export:
```
As suggested below, maybe make `certificate`, `certificate_chain` and `private_key` top level return values

diodonfrost replied:
```
RETURN = '''
-export:
-    description: Export the certificate, the certificate chain, and the encrypted private key associated with the public key embedded in the certificate
+certificate:
+    Description: The base64 PEM-encoded certificate
     returned: always
-    type: dict
-    sample:
-      certificate: XXXXXXXXXXXXXXXXXXXX
-      certificate_chain: XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
-      private_key: XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+    type: string
+certificate_chain:
+    Description: The base64 PEM-encoded certificate chain. This does not include the certificate that you are exporting
+    returned: always
+    type: string
+private_key:
+    Description: The encrypted private key associated with the public key in the certificate. The key is output in PKCS #8 format and is base64 PEM-encoded
+    returned: always
+    type: string
```

diodonfrost replied:
return `certificate`, `certificate_chain` and `private_key`

diodonfrost committed:
fix doc and issues

-   Use singular certificate name
-   Refactoring `module.exit_json`
-   Removing `argument_spec`
-   Fix doc

diodonfrost committed:
fix check failed

diodonfrost committed:
rename acm module

diodonfrost committed:
removing blank line at the end of file

diodonfrost committed:
rename file

diodonfrost committed:
re change module name
